p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -0,0 +1,396 @@
+	// that contain an NMT node (inner and leaf nodes).
+	Sha256Namespace8Flagged = 0x7701
+
+	// DagParserFormatName can be used when putting into the IPLD Dag
+	DagParserFormatName = "extended-square-row-or-col"
+
+	// FIXME: These are the same as types.ShareSize and types.NamespaceSize.
+	// Repeated here to avoid a dependency to the wrapping repo as this makes
+	// it hard to compile and use the plugin against a local ipfs version.
+	// TODO: plugins have config options; make this configurable instead
+	namespaceSize = 8
+	shareSize     = 256
+)
+
+func init() {
+	mustRegisterNamespacedCodec(
+		Sha256Namespace8Flagged,
+		"sha2-256-namespace8-flagged",
+		2*namespaceSize+sha256.Size,
+		sumSha256Namespace8Flagged,
+	)
+}
+
+func mustRegisterNamespacedCodec(
+	codec uint64,
+	name string,
+	defaulLength int,
+	hashFunc mh.HashFunc,
+) {
+	if _, ok := mh.Codes[codec]; !ok {
+		// make sure that the Codec wasn't registered from somewhere different than this plugin already:
+		if _, found := mh.Codes[codec]; found {
+			panic(fmt.Sprintf("Codec 0x%X is already present: %v", codec, mh.Codes[codec]))
+		}
+		// add to mh.Codes map first, otherwise mh.RegisterHashFunc would err:
+		mh.Codes[codec] = name
+		mh.Names[name] = codec
+		mh.DefaultLengths[codec] = defaulLength
+
+		if err := mh.RegisterHashFunc(codec, hashFunc); err != nil {
+			panic(fmt.Sprintf("could not register hash function: %v", mh.Codes[codec]))
+		}
+	}
+}
+
+// sumSha256Namespace8Flagged is the mh.HashFunc used to hash leaf and inner nodes.
+// It is registered as a mh.HashFunc in the go-multihash module.
+func sumSha256Namespace8Flagged(data []byte, _length int) ([]byte, error) {
+	isLeafData := data[0] == nmt.LeafPrefix
+	if isLeafData {
+		return nmt.Sha256Namespace8FlaggedLeaf(data[1:]), nil
+	}
+	return nmt.Sha256Namespace8FlaggedInner(data[1:]), nil
+}
+
+var _ plugin.PluginIPLD = &LazyLedgerPlugin{}
+
+type LazyLedgerPlugin struct{}
+
+func (l LazyLedgerPlugin) RegisterBlockDecoders(dec format.BlockDecoder) error {
+	dec.Register(Nmt, NmtNodeParser)
+	return nil
+}
+
+func (l LazyLedgerPlugin) RegisterInputEncParsers(iec coredag.InputEncParsers) error {
+	iec.AddParser("raw", DagParserFormatName, DataSquareRowOrColumnRawInputParser)
+	return nil
+}
+
+func (l LazyLedgerPlugin) Name() string {
+	return "LazyLedger"
+}
+
+func (l LazyLedgerPlugin) Version() string {
+	return "0.0.0"
+}
+
+func (l LazyLedgerPlugin) Init(env *plugin.Environment) error {
+	return nil
+}
+
+// DataSquareRowOrColumnRawInputParser reads the raw shares and extract the IPLD nodes from the NMT tree.
+// Note, to parse without any error the input has to be of the form:
+//
+// <share_0>| ... |<share_numOfShares>
+//
+// To determine the share and the namespace size the constants
+// types.ShareSize and types.NamespaceSize are used.
+func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int) ([]node.Node, error) {
+	const extendedSquareSize = 256
```

Comment:
Unclear variable name. This looks to be the _default_ extended square size for the purposes of reserving some memory, not the actual extended square size.

Reply:
Ah, I think you're assuming the max original square size is 128? https://github.com/lazyledger/lazyledger-core/blob/585566317e519bbb6d35d149b7e856c4c1e8657c/types/consts.go#L23

If so that's opaque and should be documented (😆 https://github.com/lazyledger/lazyledger-app/pull/21#discussion_r577435274)

Reply:
I thought the input here would be a full row or column (as the method name says). If the orig. square size is 128, then 256 is actually correct or am I missing sth?

Reply:
Also, note that while this coredag.DagParser is implemented here so this plugin can be used from
the commandline and for testing, the ipld Nodes will rather be created together with the NMT root instead of re-computing it here.
There is no reason to compute the NMT twice.

Reply:
Is this clear enough: https://github.com/lazyledger/lazyledger-core/pull/155/commits/10aa54ceef739febaf4b15b83fac94fd0e5c9d25 ?